### PR TITLE
sortable: Add Sortable component for reorderable drag-and-drop lists

### PR DESCRIPTION
**Pull request by nightscape**
## Summary

- Adds a generic `Sortable<T>` component for reorderable drag-and-drop lists
- Supports vertical and horizontal orientation via `.axis()`
- Cross-list drag-and-drop works automatically between `Sortable` instances sharing the same `T`
- Drag preview shows actual item content at original layout size
- Ghost placeholder at drag source position with faded outline
- Container-level drop targeting with centered indicator line between items
- Automatic drag state cleanup on drag cancel via `observe_release`
- `SortableState<T>` with `set_items()` for external state reconciliation (e.g. CDC streams)
- Includes `SortableStory` demo with two cross-list sortable columns

## AI Assistance

> 🤖 This implementation was pair-programmed with Claude (AI). All code has been reviewed, tested manually, and refactored to follow existing gpui-component patterns (Entity state, RenderOnce, Styled trait, builder API).

## Test plan

- [x] `cargo clippy -- --deny warnings` passes
- [x] `cargo fmt --check` passes (no changes in new files)
- [x] `typos` spell check passes
- [x] Manual testing via `cargo run` → Gallery → Sortable story
  - [x] Drag items within a list to reorder
  - [x] Drag items between lists (cross-list)
  - [x] Drop indicator line appears centered between items
  - [x] Ghost placeholder shows at original position during drag
  - [x] Drag preview matches item content and size
  - [x] Cancelling a drag (releasing outside drop zone) restores item appearance
  - [x] Only one list shows a drop indicator at a time

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/ui/src/sortable/sortable.rs [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+                }
+                let rendered = (render_fn)(item, index, window, cx);
+                container = container.child(rendered);
+            }
+            return container;
+        }
+
+        // Container-level drag_move: compute drop target from cursor position.
+        // Each item's on_drag_move reports its bounds; the container's
+        // on_drag_move uses those to determine which half the cursor is in.
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs]: Fixed — updated the comment to accurately describe the equal-division algorithm.

File: crates/ui/src/sortable/sortable.rs [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+
+                    // Compute target index by dividing container evenly among items.
+                    let target = if item_count == 0 {
+                        0
+                    } else if axis == Axis::Vertical {
+                        let relative = (cursor.y - bounds.origin.y).max(px(0.));
+                        let slot_height = bounds.size.height / item_count as f32;
+                        let slot = (relative / slot_height).floor() as usize;
+                        let within_slot = relative - slot_height * slot as f32;
+                        let insert_index = if within_slot > slot_height / 2. {
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:264]: Valid point. The equal-division approach works well for the common case (uniform items with small gaps), but will produce slightly off results with highly variable item sizes. Tracking per-item bounds via `on_prepaint` or a custom Element would fix this but adds significant complexity. Keeping the current approach for now as a pragmatic starting point — the UX is good in practice — and noted as a known limitation for a future iteration.

File: crates/ui/src/sortable/sortable.rs [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+            zone_index
+        };
+
+        state.update(cx, |s, cx| {
+            let item = s.items.remove(from);
+            s.items.insert(effective_to, item);
+            s.dragging_from_index = None;
+            s.drag_item_size = None;
+            cx.notify();
+        });
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:473]: Fixed — added `from >= state.read(cx).items.len()` guard that treats stale indices as a cancelled drag (clears state and returns). Also extracted `clear_fields()` helper to ensure `drop_target_index` is consistently cleared in all paths.

File: crates/story/src/gallery.rs [hunk 4]
```diff
@@ -81,6 +81,7 @@ impl Gallery {
                     StoryContainer::panel::<SheetStory>(window, cx),
                     StoryContainer::panel::<SidebarStory>(window, cx),
                     StoryContainer::panel::<SkeletonStory>(window, cx),
+                    StoryContainer::panel::<SortableStory>(window, cx),
                     StoryContainer::panel::<SliderStory>(window, cx),
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/story/src/gallery.rs]: Fixed — moved `SortableStory` after `SliderStory` to maintain alphabetical order.

File: crates/ui/src/sortable/sortable.rs [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+        let insert_at = zone_index.min(state.read(cx).items.len());
+        state.update(cx, |s, cx| {
+            s.items.insert(insert_at, item.clone());
+            cx.notify();
+        });
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:495]: Fixed — `clear_fields()` is now called in both source and target state updates during cross-list drops, clearing `drop_target_index` along with the other drag state.

File: crates/ui/src/sortable/sortable.rs [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    fn style(&mut self) -> &mut StyleRefinement {
+        &mut self.style
+    }
+}
+
+impl<T: Clone + 'static> RenderOnce for Sortable<T> {
+    fn render(self, window: &mut Window, cx: &mut App) -> impl IntoElement {
+        let items = self.state.read(cx).items.clone();
+        let dragging_from = self.state.read(cx).dragging_from_index;
+        let drop_target = self.state.read(cx).drop_target_index;
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:193]: Valid concern. The clone is needed because `self.state.read(cx)` borrows through `cx`, but the closures created in the loop also need `cx` (via `state.clone()` / `state.entity_id()`). Avoiding the clone would require restructuring to separate the read borrow from the closure captures, which is non-trivial with GPUI's ownership model. For typical sortable lists (tens of items, not thousands), this is fine. For very large lists, `VirtualList` should be used instead. Noting as a future optimization opportunity.

File: crates/ui/src/sortable/sortable.rs [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+fn handle_drop<T: Clone + 'static>(
+    drag: &SortableDragData<T>,
+    zone_index: usize,
+    this_list_id: gpui::EntityId,
+    state: &Entity<SortableState<T>>,
+    on_reorder: Option<&dyn Fn(usize, usize, &mut Window, &mut App)>,
+    on_insert: Option<&dyn Fn(T, usize, Entity<SortableState<T>>, &mut Window, &mut App)>,
+    window: &mut Window,
+    cx: &mut App,
+) {
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:452]: Added 6 unit tests covering: `is_noop_target` edge cases, `clear_fields` state reset, reorder index adjustment logic (forward and backward moves), full reorder simulation with vec operations, cross-list move simulation, and `SortableState` builder pattern.

File: crates/ui/src/sortable/sortable.rs [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    }
+
+    pub fn items_mut(&mut self) -> &mut Vec<T> {
+        &mut self.items
+    }
+
+    pub fn set_items(&mut self, items: Vec<T>, cx: &mut Context<Self>) {
+        self.items = items;
+        self.dragging_from_index = None;
+        self.drag_item_size = None;
```
[the remark this answers is not in the document]
Comment (nightscape) [reply crates/ui/src/sortable/sortable.rs:66]: Fixed — `set_items()` now also clears `drop_target_index` and `_drag_subscription`, fully resetting all drag state on external reconciliation.

**Review by huacnlee — CHANGES_REQUESTED** [review]
Please follow this API:

https://reui.io/patterns/sortable

1. `Sortable` should be a list container.
2. `SortableItem` is actully dragable item like the [Stepper](https://github.com/longbridge/gpui-component/blob/7e48a97a5bde0f63108c0b91a50d6b162bd68b9e/crates/ui/src/stepper/stepper.rs#L12) and [StepperItem](https://github.com/longbridge/gpui-component/blob/7e48a97a5bde0f63108c0b91a50d6b162bd68b9e/crates/ui/src/stepper/item.rs#L14)
3. Story use normal style like the stories in Reui.
4. I expected there is no `SortaleState`, it's very simple drag drop case, can we just have Sortable > SortableItem as stateless? Like the Stepper

**Comment by Trojanking123** [comment]
this is useful for me. hope to be merged soon.

**Comment by MSIsunny** [comment]
looks good for me.

**Comment by huacnlee** [comment]
Close too long no response PRs.
